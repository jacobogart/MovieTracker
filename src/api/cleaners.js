import { imageUrl } from '../utils/API-logins';

export const cleanAllMovies = (moviesArray) => {
  const cleanMovies = moviesArray.map(movie => {
    const { 
      id, 
      title, 
      poster_path, 
      overview, 
      release_date, 
      vote_average 
    } = movie;

    return { 
      id, 
      title, 
      posterImg: `${imageUrl}${poster_path}`, 
      overview, 
      releaseDate: release_date, 
      rating: vote_average 
    };
  });

  return cleanMovies;
};

export const cleanPeople = (data) => {
  const director = data.crew.filter(person => person.job === "Director");
  const writer = data.crew.filter(person => person.job === "Writer");
  const cast = data.cast.map(person => {
    let { name, character, profile_path,  cast_id} = person;

    return { 
      name, 
      character, 
      headshot: `${imageUrl}${profile_path}`, 
      id: cast_id 
    };
  });

  return { director, writer, cast };
};