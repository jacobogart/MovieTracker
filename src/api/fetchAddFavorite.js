import { serverUrl } from "./pathNames";

export const fetchAddFavorite = (cardData) => {
  const url = `${serverUrl}/favorites/new`;
  const { 
    user, 
    title, 
    overview, 
    posterImg, 
    id, 
    rating, 
    releaseDate 
  } = cardData;
  let poster_path = posterImg.split('500')[1];
  const body = {
    movie_id: id, 
    user_id: user.id,
    title, 
    poster_path,
    release_date: releaseDate, 
    vote_average: rating, 
    overview
  };
  const init = {
    method: 'POST',
    body: JSON.stringify(body),
    headers: { 'Content-Type': 'application/json' }
  };

  return fetch(url, init)
    .then(response => {
      if (!response.ok) {
        throw Error('Failed to add favorite');
      } else {
        return response.json();
      }
    });
};
