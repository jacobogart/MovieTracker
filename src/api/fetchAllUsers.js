import { serverUrl } from "./pathNames";

export const fetchAllUsers = () => {
  return fetch(serverUrl)
    .then(response => {
      if (!response.ok) {
        throw new Error("Failed to create account");
      } else {
        return response.json();
      }
    })
    .then(data => data.data);
};
