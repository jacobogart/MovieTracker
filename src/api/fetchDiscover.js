import { jacobKey, baseUrl } from '../utils/API-logins'

const fetchDiscover = () => {
  return fetch(`${baseUrl}discover/movie${jacobKey}`)
    .then(response => {
      if (!response.ok) {
        throw new Error('Discover failed to fetch');
      } else {
        return response.json();
      }
  });
}

export default fetchDiscover;

