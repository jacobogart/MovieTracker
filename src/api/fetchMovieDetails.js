import { jacobKey } from '../utils/API-logins';
import { baseUrl } from './pathNames';

const fetchMoviePeople = (id) => {
  return fetch(`${baseUrl}movie/${id}/credits${jacobKey}`)
    .then(response => {
      if (!response.ok) {
        throw new Error('People failed to fetch');
      } else {
        return response.json();
      }
    });
};

export default fetchMoviePeople;
