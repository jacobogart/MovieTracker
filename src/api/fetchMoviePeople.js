import { jacobKey } from '../utils/API-logins';
import { baseUrl } from './pathNames';

const fetchMoviePeople = (id) => {
  return fetch(`${baseUrl}movie/${id}/credits${jacobKey}`)
    .then(response => {
      if (!response.ok) {
        throw new Error('Failed to fetch movie cast and crew');
      } else {
        return response.json();
      }
    });
};

export default fetchMoviePeople;
