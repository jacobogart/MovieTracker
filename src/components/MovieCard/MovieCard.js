import React from 'react';
import { Link } from 'react-router-dom';

export const MovieCard = ({ title, overview, posterImg, id, rating, releaseDate }) => {
  let movieOverview;
  let ratingNum;
  let releaseYear = releaseDate.substring(0, 4);
  let percentage = rating * 10;

  if (overview.length > 150) {
    let shortOverview = overview
      .split(' ')
      .slice(0, 35)
      .join(' ');

    movieOverview = `${shortOverview}...`;
  } else {
    movieOverview = overview;
  }

  if (rating.toString().length === 1) {
    ratingNum = `${rating}.0`;
  } else {
    ratingNum = rating;
  }

  let ratingCircle = 
    <svg viewBox="0 0 36 36" className="rating-circle">
      <path className="circle-bg"
        d="M18 2.0845
          a 15.9155 15.9155 0 0 1 0 31.831
          a 15.9155 15.9155 0 0 1 0 -31.831"/>
      <path className="circle"
        strokeDasharray={`${percentage}, 100`}
        d="M18 2.0845
          a 15.9155 15.9155 0 0 1 0 31.831
          a 15.9155 15.9155 0 0 1 0 -31.831"/>
      <text x="9" y="23" className="rating-num">{ratingNum}</text>
    </svg>

  return (
    <article className="MovieCard">
      <img className="card-img" src={posterImg} alt={`${title} poster`}/>
      <div className="card-info">
        <div className="card-header">
          <div className="circle-wrapper">
            {ratingCircle}
          </div>
          <div className="card-title-wrap">
            <h2 className="card-title">{title}</h2>
            <p className="card-year">({releaseYear})</p>
          </div>
        </div>
        <p className="card-overview">{movieOverview}</p>
        <Link 
          className="more-info-btn" 
          role="button"
          to={`/movie/${id}`}
        >
          More Info
        </Link>
      </div>
    </article>
  );
} 



export default MovieCard;