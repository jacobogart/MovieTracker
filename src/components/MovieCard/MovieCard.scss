.MovieCard {
  box-shadow: $box-shadow;
  display: flex;
  margin: 10px;
  text-align: left;
  width: 525px;
  height: 300px;

  .card-img {
    height: 300px;
    object-fit: cover;
  }

  .card-info {
    padding: 20px;
    position: relative;

    .card-header {
      display: flex;

      .rating-circle {
        border-radius: 50%;
        width: 40px;
        height: 40px;
      }
    
      .circle-bg {
        fill: none;
        stroke: #999;
        stroke-width: 4px;
      }

      .circle {
        fill: none;
        stroke: #4CC790;
        stroke-width: 4px;
        stroke-linecap: round;
        animation: progress 1s ease-out forwards;
      }

      .rating-num {
        font-size: .8em;
      }

      .card-title-wrap {
        margin: 0 0 0 10px;

        .card-title {
          font-size: 1.1em;
          margin: 0;
        }

        .card-year {
          margin: 0;
        }
      }
    }

    .card-overview {
      height: 120px;
      overflow: auto;
      margin: 20px 0;
    }

    .more-info-btn {
      background: black;
      border-top: 3px solid black;
      bottom: 0;
      color: white;
      font-weight: 600;
      left: 0;
      padding: 15px;
      position: absolute;
      right: 0;
      text-align: center;

      &:hover {
        background: #333;
      }
    }
  }
}

@keyframes progress {
  0% {
    stroke-dasharray: 0 100;
  }
}