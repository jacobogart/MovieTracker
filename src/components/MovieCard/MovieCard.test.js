import { MovieCard } from "./MovieCard";
import { shallow } from "enzyme";
import React from "react";
import { cleanMovie, mockUser } from "../../api/__tests__/mockData";
import { fetchAddFavorite } from "../../api/fetchAddFavorite";

jest.mock("../../api/fetchAddFavorite");

const mockUpdateUserFavorites = jest.fn();
const mockDeleteUserFavorite = jest.fn();
const defaultState = {
  favorites: null,
  isFavorite: false,
  error: ""
};
const userlessComponent = (
  <MovieCard
    {...cleanMovie}
    user={{}}
    updateUserFavorites={mockUpdateUserFavorites}
    deleteUserFavorite={mockDeleteUserFavorite}
    key={111}
  />
);
const userComponent = (
  <MovieCard
    {...cleanMovie}
    user={mockUser}
    updateUserFavorites={mockUpdateUserFavorites}
    deleteUserFavorite={mockDeleteUserFavorite}
    key={111}
  />
);

describe("MovieCard", () => {
  it("should match the snapshot", () => {
    let wrapper = shallow(userComponent);
    expect(wrapper).toMatchSnapshot();
  });

  it('should have a default state', () => {
    let wrapper = shallow(userlessComponent, { disableLifecycleMethods: true });
    expect(wrapper.state()).toEqual(defaultState);
  });

  describe('handleFavorite', () => {
    it('should invoke fetchAddFavorite if isFavorite is false and the user is logged in', () => {
      let wrapper = shallow(userComponent);
      fetchAddFavorite.mockImplementation(() => Promise.resolve())
      wrapper.instance().handleFavorite();
      expect(fetchAddFavorite).toHaveBeenCalled();
    });

    it('should invoke updateUserFavorites after fetchAddFavorite', () => {
      let wrapper = shallow(userComponent);
      fetchAddFavorite.mockImplementation(() => Promise.resolve());
      wrapper.instance().handleFavorite();
      expect(mockUpdateUserFavorites).toHaveBeenCalled();
    });

    it.skip('should set state to an error if the fetch fails', () => {
      let wrapper = shallow(userComponent);
      fetchAddFavorite.mockImplementation(() => {throw new Error('error')});
      wrapper.instance().handleFavorite();
      expect(wrapper.state('error')).toEqual('error')
    });

    it('should invoke deleteUserFavorite if the movie is already favorited', () => {
      let wrapper = shallow(userComponent);
      fetchAddFavorite.mockImplementation(() => Promise.resolve());
      wrapper.instance().handleFavorite();
      expect(mockUpdateUserFavorites).toHaveBeenCalled();
    });

    it.skip('should prompt the user to login if they favoroite without being logged in', () => {
      let wrapper = shallow(userlessComponent);
      // expect something to have been called
    });
  });
});
