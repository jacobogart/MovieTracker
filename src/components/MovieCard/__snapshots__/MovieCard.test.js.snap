// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard should match the snapshot 1`] = `
<article
  className="MovieCard"
>
  <img
    alt="Avengers: Endgame poster"
    className="card-img"
    src="https://image.tmdb.org/t/p/w500/or06FN3Dka5tukK1e9sl16pB3iy.jpg"
  />
  <div
    className="card-info"
  >
    <div
      className="card-header"
    >
      <div
        className="card-header-left"
      >
        <div
          className="circle-wrapper"
        >
          <svg
            className="rating-circle"
            viewBox="0 0 36 36"
          >
            <path
              className="circle-bg"
              d="M18 2.0845 a 15.9155 15.9155 0 0 1 0 31.831 a 15.9155 15.9155 0 0 1 0 -31.831"
            />
            <path
              className="circle circle-green"
              d="M18 2.0845 a 15.9155 15.9155 0 0 1 0 31.831 a 15.9155 15.9155 0 0 1 0 -31.831"
              strokeDasharray="86, 100"
            />
            <text
              className="rating-num"
              x="9"
              y="23"
            >
              8.6
            </text>
          </svg>
        </div>
        <div
          className="card-title-wrap"
        >
          <h2
            className="card-title"
          >
            Avengers: Endgame
          </h2>
          <p
            className="card-year"
          >
            (
            2019
            )
          </p>
        </div>
      </div>
      <div
        className="icon-wrap"
      >
        <i
          className="far fa-heart"
          onClick={[Function]}
        />
        <i
          className="far fa-bookmark"
        />
      </div>
    </div>
    <p
      className="card-overview"
    >
      After the devastating events of Avengers: Infinity War, the universe is in ruins due to the efforts of the Mad Titan, Thanos. With the help of remaining allies, the Avengers must assemble once more in...
    </p>
    <Link
      className="more-info-btn"
      role="button"
      to="/movie/299534"
    >
      More Info
    </Link>
  </div>
</article>
`;
